llpc/lower/llpcSpirvLowerGlobal.cpp

```diff
@@ -343,26 +343,42 @@ void SpirvLowerGlobal::handleCallInst(bool checkEmitCall, bool checkInterpCall)
             auxInterpValue = callInst->getArgOperand(1); // Vertex no.
           }
 
-          if (isa<GetElementPtrInst>(loadSrc)) {
+          if (auto getElemPtr = dyn_cast<GetElementPtrInst>(loadSrc)) {
             // The interpolant is an element of the input
-            interpolateInputElement(interpLoc, auxInterpValue, *callInst);
+            SmallVector<Value *, 6> indexOperands;
+            for (auto &index : getElemPtr->indices())
+              indexOperands.push_back(m_builder->CreateZExtOrTrunc(index, m_builder->getInt32Ty()));
+            GlobalVariable *gv = cast<GlobalVariable>(getElemPtr->getPointerOperand());
+            interpolateInputElement(interpLoc, auxInterpValue, *callInst, indexOperands, gv);
           } else {
             // The interpolant is an input
             assert(isa<GlobalVariable>(loadSrc));
 
             auto input = cast<GlobalVariable>(loadSrc);
             auto inputTy = input->getValueType();
+            Type *funcRetType = callInst->getFunctionType()->getReturnType();
 
-            MDNode *metaNode = input->getMetadata(gSPIRVMD::InOut);
-            assert(metaNode);
-            auto inputMeta = mdconst::dyn_extract<Constant>(metaNode->getOperand(0));
+            if (funcRetType == inputTy) {
 
-            m_builder->SetInsertPoint(callInst);
-            auto loadValue = addCallInstForInOutImport(inputTy, SPIRAS_Input, inputMeta, nullptr, 0, nullptr, nullptr,
-                                                       interpLoc, auxInterpValue, false);
+              MDNode *metaNode = input->getMetadata(gSPIRVMD::InOut);
+              assert(metaNode);
+              auto inputMeta = mdconst::dyn_extract<Constant>(metaNode->getOperand(0));
 
-            m_interpCalls.insert(callInst);
-            callInst->replaceAllUsesWith(loadValue);
+              m_builder->SetInsertPoint(callInst);
+              auto loadValue = addCallInstForInOutImport(inputTy, SPIRAS_Input, inputMeta, nullptr, 0, nullptr, nullptr,
+                                                         interpLoc, auxInterpValue, false);
+
+              m_interpCalls.insert(callInst);
+              callInst->replaceAllUsesWith(loadValue);
+            } else {
+              // Type of GlobalVariable and return type of the Call are not the same. This can happen for opaque
+              // pointers because zero-index GEP instructions are removed. In this case we have to build zero-index
+              // vector which is equivalent to zero-index GEP and use "interpolateInputElement".
+              SmallVector<Value *, 6> indexOperands;
+              indexOperands.push_back(m_builder->getInt32(0));
+              appendZeroIndexToMatchTypes(m_builder, indexOperands, funcRetType, inputTy);
+              interpolateInputElement(interpLoc, auxInterpValue, *callInst, indexOperands, input);
+            }
```

Comment:
In the spirit of leaving the code in a slightly better state than we found it in, please refactor this into a code sequence that calls `interpolateInputElement` in only one place (unified for the GEP and non-GEP case) and never calls `addCallInstForInOutImport` directly (let `interpolateInputElement` handle that for us).